pandas/io/orc.py

```diff
@@ -52,3 +58,106 @@ def read_orc(
     with get_handle(path, "rb", is_text=False) as handles:
         orc_file = orc.ORCFile(handles.handle)
         return orc_file.read(columns=columns, **kwargs).to_pandas()
+
+
+def to_orc(
+    df: DataFrame,
+    path: FilePath | WriteBuffer[bytes] | None = None,
+    *,
+    engine: Literal["pyarrow"] = "pyarrow",
+    index: bool | None = None,
+    **kwargs,
+) -> bytes | None:
+    """
+    Write a DataFrame to the ORC format.
+
+    .. versionadded:: 1.5.0
+
+    Parameters
+    ----------
+    df : DataFrame
+        The dataframe to be written to ORC. Raises NotImplementedError
+        if dtype of one or more columns is category, unsigned integers,
+        intervals, periods or sparse.
+    path : str, file-like object or None, default None
+        If a string, it will be used as Root Directory path
+        when writing a partitioned dataset. By file-like object,
+        we refer to objects with a write() method, such as a file handle
+        (e.g. via builtin open function). If path is None,
+        a bytes object is returned.
+    engine : {{'pyarrow'}}, default 'pyarrow'
+        ORC library to use, or library it self, checked with 'pyarrow' name
+        and version >= 7.0.0. Raises ValueError if it is anything but
+        'pyarrow'.
+    index : bool, optional
+        If ``True``, include the dataframe's index(es) in the file output. If
+        ``False``, they will not be written to the file.
+        If ``None``, similar to ``infer`` the dataframe's index(es)
+        will be saved. However, instead of being saved as values,
+        the RangeIndex will be stored as a range in the metadata so it
+        doesn't require much space and is faster. Other indexes will
+        be included as columns in the file output.
+    **kwargs
+        Additional keyword arguments passed to the engine.
+
+    Returns
+    -------
+    bytes if no path argument is provided else None
+
+    Raises
+    ------
+    NotImplementedError
+        Dtype of one or more columns is category, unsigned integers, interval,
+        period or sparse.
+    ValueError
+        engine is not pyarrow.
+
+    Notes
+    -----
+    * Before using this function you should read the
+      :ref:`user guide about ORC <io.orc>` and
+      :ref:`install optional dependencies <install.warn_orc>`.
+    * This function requires `pyarrow <https://arrow.apache.org/docs/python/>`_
+      library.
+    * Category, unsigned integers, interval, period and sparse Dtypes
+      are not supported yet.
+    * Currently timezones in datetime columns are not preserved when a
+      dataframe is converted into ORC files.
+    """
+    if index is None:
+        index = df.index.names[0] is not None
+
+    # If unsupported dtypes are found raise NotImplementedError
+    for dtype in df.dtypes:
+        dtype_str = dtype.__str__().lower()
+        if (
```

Comment:
Will pyarrow raise if these dtypes are passed? If so, can a a pyarrow error be caught and reraised as a `NotImplementedError` so this can be more flexible to other potential dtypes not supported in the future?

Reply:
@mroeschke  It seg faults out. I need to catch them in Arrow 9.0.0. Meanwhile can we use the current dtype filter?

Reply:
Okay, this is fine then given:

1. Could you use the type checking functions in `pandas.core.dtypes.common` instead? e.g. `is_categorical_dtype(dtype)`?
2. Could you make a note that in pyarrow 9.0.0 this checking should not be needed?

Reply:
Sure!

Reply:
Done!

Since for sparse dtypes we get a `TypeError` from Arrow when converting the dataframe to a pyarrow table I plan to use `TypeError` for the other 4 in pyarrow 9.0.0 as well. The try-except block has been added.